telosnetwork/teloscan
# #503 | EIP-1967 Proxy contracts interface

# Fixes #503 

## Description
This PR enables users to interface with proxy contracts in the Contract tab of the address page.

In order to test this functionality, I have deployed some contracts on mainnet:
base contract: [`0x86A2fA5eA80A831A935ED2fAF3B3ad2F11a04Ff6`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x86A2fA5eA80A831A935ED2fAF3B3ad2F11a04Ff6#contract)
proxy contract with no methods: [`0x0cB032Af8Ff931647eD838856af731a6FBc68F7D`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x0cB032Af8Ff931647eD838856af731a6FBc68F7D#contract)
proxy contract with methods: [`0xbbd1170a77F9eA0458872AF740204e7C8f9391f1`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xbbd1170a77F9eA0458872AF740204e7C8f9391f1#contract)

base contract with overloaded methods: [`0xA42dCbe6EDdACd25BEb54B5cC8CBf273c14B5898`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xA42dCbe6EDdACd25BEb54B5cC8CBf273c14B5898#contract)
unverified proxy contract: [`0x8A44f083b1BA3112C776EF7194F572977c684c37`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x8A44f083b1BA3112C776EF7194F572977c684c37#contract)
proxy for contract with overloads: [`0xAE21A25fd0aA5e7a366ba42EfC65685c0acDdC53`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xAE21A25fd0aA5e7a366ba42EfC65685c0acDdC53#contract)
proxy with updateable implementation: [`0x359CC3AA5134379D296c369DE7fc417C6C4e3e5B`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x359CC3AA5134379D296c369DE7fc417C6C4e3e5B#contract)

**Update**: I misunderstood how proxy and implementation contracts interact when i first created this PR. I have updated the testing notes. When you call a method on an implementation contract through a proxy (from the `Read as Proxy` or `Write as Proxy` tab), the data is stored **on the proxy**. So when you call `increment` **on the proxy**, it does **not** update the `count` on the implementation base contract. the implementation contract maintains its own state which is not shared with the proxy contract. So when you call `count()` on the proxy contract using the `Read as Proxy` tab, you are not reading the implementation contract's state! You are actually reading state stored in the proxy contract in a special implementation slot. The proxy contract also "has its own state", which is why `incrementProxyCount` and `getProxyCount` (in the `Read` and `Write` tabs) do not affect the state called `count`.


## Test scenarios
- go to https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/
- log in with metamask
- go to the address page for a proxy contract whose implementation contract is _unverified_, such as https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xb99C43d3bce4c8Ad9B95a4A178B04a7391b2a6EB#contract
    - you should see 2 new tabs, 'Read as Proxy' and 'Write as Proxy'
- go to the read or write tab for the contract given above (note that for this contract, there are no executable methods)
    - you should see the message "No executable functions found in the contract ABI"
- go to the 'Read as Proxy' tab
    - you should see a message saying that "This contract is an EIP-1967 Transparent Proxy for the contract 0x1701D9B92BAf8C5d01F97CA97b328Cc4db945E93"
    - you should see another message saying "The contract for which this is a proxy is not verified, so its ABI is not available. To interact with the implementation contract, visit the implementation contract page"
    - the links should go to the expected page
- go to 'Write as Proxy' tab
    - you should see the same thing as 'Read as Proxy' tab
- go to the address page of a proxy contract whose implementation contract _is verified_, e.g. https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xbbd1170a77F9eA0458872AF740204e7C8f9391f1#contract
- go to the 'Read' tab
    - you should see some executable methods, which are exposed by the contract
- execute `getProxyCount` and take note of the number
    - this number is tracked by the contract itself, NOT on the implementation contract, i.e. this proxy contract has its own methods in addition to being a proxy
- go to the 'Write' tab
- execute `incrementProxyCount`
- go back to the 'Read' tab
- execute `getProxyCount`
    - the result should be 1 greater than the last time you called it
- go to 'Read as Proxy' tab
- execute `count`, which is a count variable stored by the _implementation_ contract
- in a new tab, go to the implementation contract page, https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x86A2fA5eA80A831A935ED2fAF3B3ad2F11a04Ff6#contract
- go to the 'Read' tab
- execute `count` and take note of the result
- go back to the proxy contract page
- go to the 'Write as Proxy' tab
- execute `increment`
- go back to 'Read as Proxy' tab
- execute `count`
    - the count should have incremented
- go back to your tab with the implementation contract
- execute `count`
    - the count should **not have changed**
- on the implementation contract page, go to the 'Write' tab
- execute `increment`
- go to the 'Write' tab
- execute `count` and take note of the result
- go back to your tab with the proxy contract
- go to the 'Read' tab
- execute `count`
    - the count should **not have changed** from the last time you looked at the proxy count

## Checklist:
